lib/doi_search.rb

```diff
@@ -1,15 +1,41 @@
 class DoiSearch
+  # The first authoritative result(s) are returned.
+  # If an authoritative local match is found, no remote services are hit.
+  # WoS hits need to be post-processed because the API does partial string
+  # matching for DOI queries, unfortunately.
+  # @param [String] doi
+  # @return [Array<Publication, Hash>] matching pubs and/or SW hashes
   def self.search(doi)
-    results = [Publication.find_by_doi(doi)].compact
-    results += ScienceWireClient.new.get_pub_by_doi(doi) if results.none?(&:authoritative_doi_source?)
-    results.reject! { |pub| non_sw_pub pub } if results.size > 1
+    pub = Publication.find_by_doi(doi)
+    return [pub] if pub && pub.authoritative_doi_source?
+    if Settings.SCIENCEWIRE.enabled
+      sw_hits = ScienceWireClient.new.get_pub_by_doi(doi)
+      return sw_hits unless sw_hits.empty?
+    end
+    results = []
+    if Settings.WOS.enabled
+      results += WebOfScience.queries.search_by_doi(doi)
+      full_match = normalized_dois(results)[normalized_doi(doi)]
+      return [full_match] if full_match
+    end
+    # Everything else is non-authoritative
+    results.unshift(pub) if pub
     results
   end
```

Comment:
:+1: for use of Settings
- the cognitive complexity has gone up :hankey: but punting on this for now
ln 9-10 - seems to behave the same as the old code
ln 10 - if pub is not authoritative (manual pub), the pub hangs around for the rest of the method (just noting behavior)
ln 11-14 makes sense although it's not obvious that `sw_hits` is an array (the old code ensured it and make it clearer)
ln 15-20 and called methods are complex, but OK;  :+1: for `normalized_doi`
  - the Hash construction is complex and hits all the results (it's difficult to read), maybe an `Array.find` would be easier to read? e.g. `results.find {|i| normalized_doi(i) == normalized_doi(doi) }` but do the `normalized_doi(doi)` first (not every time);  assuming the results are not very large, the trade-off for easier to read might be more important than performance of Hash constuct vs. the array find (?)

Reply:
I had Aaron spot-check the `Hash[]` constructor line asking if it was readable, and he was fine w/ it.

Reply:
@darrenleeweber Yes, you're right about the method signature, will update that to include `WebOfScience::Record`

Reply:
Line 22 is informed by the above comment.  If we reached the bottom of the method, we have no authoritative matches.  We might have `pub` and might have some non-exact WOS matches.  The response will include both when available, with the local one first.